c7n/credentials.py

```diff
@@ -56,8 +57,15 @@ def __call__(self, assume=True, region=None):
 
         session._session.user_agent_name = self.user_agent_name
         session._session.user_agent_version = version
+
+        for s in self._subscribers:
+            s(session)
+
         return session
 
+    def set_subscribers(self, subscribers):
```

Comment:
subscriber needs tests, also should consider if we want a variant on this api, as it effectively is single invoker (vs say subscribers.add/iter/clear)